src/App.js

```diff
@@ -97,105 +115,49 @@ let App = () => {
-            eventsForModule.splice(i, 1)
-          }
-        }
-      }
-      // Delete hidden occurrences
-      for (const [module, groupId, occurrence] of hiddenOccurrences) {
-        if (module !== id) continue
-        for (let i = eventsForModule.length - 1; i >= 0; i--) {
-          const event = eventsForModule[i]
-          if (event.activity === groupId && parseInt(event.occurrence) === occurrence) {
-            eventsForModule.splice(i, 1)
-          }
-        }
-      }
-
-      // Add currently visible events to the calendar
-      api.addEventSource({
-        id,
-        color: stringToColor(id),
-        events: parseEvents(eventsForModule, year, session, id)
-      })
-    })
-  }, [timetableData, year, session, selectedModules, calendar, timeZone, m, modules, specifiedOccurrences, hiddenOccurrences])
 
-  // Remove specified events for modules that have been removed
-  useEffect(() => {
-    updateSpecifiedOccurrences()
-    updateHiddenOccurrences()
-  // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, [selectedModules])
-
-  // We select occurrences by adding them to specifiedOccurrences, which
-  // edits the query string in an effect
-  const selectOccurrence = (module, groupId, occurrence) => {
-    // Eg adding ['COMP1130', 'ComA', 1]
-    setSpecifiedOccurrences([...specifiedOccurrences, [module, groupId, occurrence]])
-  }
-  const resetOccurrence = (module, groupId, occurrence) => {
-    setSpecifiedOccurrences(specifiedOccurrences.filter(
-      ([m, g, o]) => !(m === module && g === groupId && o === occurrence)
-    ))
+  const changeOccurrences = (state, action) => {
+    let [module, groupId, occurrence] = action.values
+    switch (action.type) {
+      case 'select':
+        setQueryParam(module, groupId + occurrence)
+        return [...state, action.values]
+      case 'reset':
+        setQueryParam(module, '')
+        return state.filter(
+          ([m, g, o]) => !(m === module && g === groupId && o === occurrence)
+        )
+      default:
+        throw new Error()
+    }
   }
-  const hideOccurrence = (module, groupId, occurrence) => {
-    setHiddenOccurrences([...hiddenOccurrences, [module, groupId, occurrence]])
+
+  const [specifiedOccurrences, setSpecifiedOccurrences] = useReducer(changeOccurrences, getSpecOccurrences())
+
+  const changeHidden = (state, action) => {
+    switch (action.type) {
+      case 'reset':
+        unsetQueryParam('hide')
+        return []
+      case 'hide':
+        // Should we have a hide url parameter
+        const new_state = [...state, action.values]
+        const hide = new_state.map(x => x.join('_')).join(',')
+        if (hide.length > 0)
+          setQueryParam('hide', hide)
+        else
+          unsetQueryParam('hide')
+        return new_state
+      default:
+        throw new Error()
+    }
   }
+  // Events that are manually hidden with the eye icon
+  const [hidden, setHidden] = useReducer(changeHidden, h)
```

Comment:
This was `hiddenOccurrences` because we also have `hiddenDays`, I reckon keep the old name until we've refactored `hiddenDays` into a different scope